maintainers/scripts/instantiate-constitutents.sh: init

amjoseph-nixpkgs opened the pull request:
This commit adds a script which lets you "do what Hydra does" using only `nix`.  See the comment at the top of the file for sample usage.

This script relies on `make -j` to run large numbers of `nix-instantiate`s in parallel.  You'll need to tune the `-j` value appropriately for your machine's core/memory balance.  On my 32-core opteron machines I find that -j64 keeps all the cores busy, using about 18GB of memory (~250MB per `nix-instance`).

Commit:
maintainers/scripts/instantiate-constitutents.sh: init

This commit adds a script which lets you "do what Hydra does" using
only `nix`.  See the comment at the top of the file for sample
usage.

This script relies on `make -j` to run large numbers of
`nix-instantiate`s in parallel.  You'll need to tune the `-j` value
appropriately for your machine's core/memory balance.  On my 32-core
opteron machines I find that -j64 keeps all the cores busy, using
about 18GB of memory (~250MB per `nix-instance`).

Co-authored-by: Bernardo Meurer <bernardo@meurer.org>

lovesegfault commented:
Isn't this basically @Mic92's `nix-eval-jobs`?

https://github.com/nix-community/nix-eval-jobs

amjoseph-nixpkgs commented:
> Isn't this basically @Mic92's `nix-eval-jobs`?

I don't think so.  `nix-eval-jobs` doesn't seem to know how to turn a Hydra job's `.constituents` into a pile of separate derivations, which is what this script does.  Instead it just gives you a single derivation which doesn't actually build the constituents:

```
nix-shell -p nix-eval-jobs --command 'nix-eval-jobs --expr "(import <nixpkgs/nixos/release-combined.nix> { supportedSystems = [\"x86_64-linux\"]; }).tested"'
warning: `--gc-roots-dir' not specified
{"attr":"","attrPath":[],"drvPath":"/nix/store/iraxp77m9qgkq8xd4dyabj36n6i94vis-nixos-23.05pre56789.gfedcba.drv","name":"nixos-23.05pre56789.gfedcba","outputs":{"out":"/nix/store/5zgj84nhm8yciyx2mimf9vw5lb6d1m78-nixos-23.05pre56789.gfedcba"},"system":"x86_64-linux"}
```

If you run `nix show-derivation` on the derivation above you'll see that it has `constituents` as an environment variable.  That derivation won't build the constituents; all it does is write their names into a file.

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
```
```suggestion
#!/usr/bin/env bash
```

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,31 @@
+#!/bin/sh
+
```
```suggestion
set -euo pipefail
```

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,31 @@
+# Example:
+#   make -j32 -f <(instantiate-constitutents.sh '<nixpkgs/nixos/release-small.nix>' tested 'supportedSystems = ["x86_64-linux"];')
+
+DEST="$(mktemp -d)"
+echo "drv files will be written to $DEST" 1>&2
+IMPORT="$1"
+shift
+ATTR="$1"
+shift
+ARGS="$1"
```
```suggestion
if [ "$#" -neq 3 ]; then
    cat <<-EOF 
    This script expects three arguments:
    1. A nix expression to instantiate (e.g. `<nixpkgs/nixos/release-small.nix>`)
    2. The name of the Hydra jobset attr within (1) (e.g. `tested`)
    3. Additional args to `nix-instantiate` (e.g. `system = ["aarch64-linux"]'`
    
    Example usage:
    $ make -j32 -f <(instantiate-constitutents.sh '<nixpkgs/nixos/release-small.nix>' tested 'supportedSystems = ["x86_64-linux"];')
    EOF
    exit 1
fi
DEST="$(mktemp -d)"
echo "drv files will be written to $DEST" 1>&2
IMPORT="$1"
ATTR="$2"
ARGS="$3"

readonly DEST IMPORT ATTR ARGS
```

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,31 @@
+ARGS="$1"
+
+echo "default: build"
+for ATTR in $(nix eval --raw --impure --expr "let lib = import <nixpkgs/lib>; in lib.concatStringsSep \"\\n\" (with import $IMPORT {$ARGS}; $ATTR.constituents)")
```
Move the eval outside the `for ... in` and into a variable

amjoseph-nixpkgs replied:
2002a491c657

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,31 @@
+echo "default: build"
+for ATTR in $(nix eval --raw --impure --expr "let lib = import <nixpkgs/lib>; in lib.concatStringsSep \"\\n\" (with import $IMPORT {$ARGS}; $ATTR.constituents)")
+do \
+    echo "$DEST/$ATTR:; @nix-instantiate --expr '(import $IMPORT {$ARGS}).$ATTR' --add-root $DEST/$ATTR"
+    echo "drvs: $DEST/$ATTR"
+    echo "build: $DEST/$ATTR"
+    echo "DRVS += $DEST/$ATTR"
```
IMO use a heredoc to make this nicer:

```bash
cat <<-EOF
    ...
EOF
```

amjoseph-nixpkgs replied:
682f69d5dea8

lovesegfault commented on maintainers/scripts/instantiate-constitutents.sh:
Make sure to run `shellcheck` and `shfmt` on this too

amjoseph-nixpkgs replied:
`shfmt`: 1d19fdf1302b

amjoseph-nixpkgs replied:
`shellcheck` passes, except that it thinks the here-doc is bash code.  Deleting the body of the here-doc causes it to pass.

amjoseph-nixpkgs commented:
Note: I need to extend this to allow multiple attributes.  Draft until that is done.

amjoseph-nixpkgs commented:
Squashed, rebased.  Currently in use to test https://github.com/NixOS/nixpkgs/pull/237856 [here](https://github.com/NixOS/nixpkgs/pull/237856#issuecomment-1633466188)

ncfavier commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,44 @@
+ATTRS=$(nix eval --raw --impure --expr "let lib = import <nixpkgs/lib>; in lib.concatStringsSep \"\\n\" (with import $IMPORT {$ARGS}; $TOPLEVEL_ATTR.constituents)")
+for ATTR in $ATTRS; do
+    cat << EOF
+$DEST/$ATTR:; @nix-instantiate --expr '(import $IMPORT {$ARGS}).$ATTR' --add-root $DEST/$ATTR
```
```suggestion
$DEST/$ATTR:; @nix-instantiate --expr '(import '${IMPORT@Q}' {'${ARGS@Q}'}).'${ATTR@Q} --add-root $DEST/$ATTR
```

Bare minimum for safe injection into single quotes. You may also want to consider whether to inject `import` and `attr` as Nix code or pass them via `--argstr import ${import@Q}`

ncfavier replied:
I actually don't know if `@Q` works on macos, their bash is pretty old. Needs double checking

amjoseph-nixpkgs commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,44 @@
+#   make -j32 -f <(instantiate-constitutents.sh '<nixpkgs/nixos/release-small.nix>' tested 'supportedSystems = ["x86_64-linux"];')
+
+if [ ! "$#" -eq 3 ]; then
+    cat << EOF
```
Not sure why this matters but my `bash`-fu is probably not as strong as yours :)

ncfavier replied:
Makes the heredoc be treated literally. Otherwise the backticks would be interpreted as command substitutions

amjoseph-nixpkgs replied:
Oh, I see, it's the same as `\EOF`.

Commit:
Update maintainers/scripts/instantiate-constitutents.sh

Co-authored-by: Naïm Favier <n@monade.li>

Commit:
Update maintainers/scripts/instantiate-constitutents.sh

Co-authored-by: Naïm Favier <n@monade.li>

amjoseph-nixpkgs commented on maintainers/scripts/instantiate-constitutents.sh:
```diff
@@ -0,0 +1,44 @@
+EOF
+    exit 1
+fi
+DEST="$(mktemp -d)"
+echo "drv files will be written to $DEST" 1>&2
+IMPORT="$1"
+TOPLEVEL_ATTR="$2"
+ARGS="$3"
```
fc6d04f410d6
